Sources/X509/CustomPrivateKey.swift

```diff
@@ -0,0 +1,73 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftCertificates open source project
+//
+// Copyright (c) 2025 Apple Inc. and the SwiftCertificates project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftCertificates project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+#if canImport(FoundationEssentials)
+import FoundationEssentials
+#else
+import Foundation
+#endif
+import SwiftASN1
+
+@available(macOS 10.15, iOS 13, watchOS 6, tvOS 13, macCatalyst 13, visionOS 1.0, *)
+public protocol CustomPrivateKey: Sendable, Hashable, PEMSerializable {
+
+    /// Obtain the ``Certificate/PublicKey-swift.struct`` corresponding to
+    /// this private key.
+    var publicKey: Certificate.PublicKey { get }
+
+    var defaultSignatureAlgorithm: Certificate.SignatureAlgorithm { get }
+
+    /// Return a list of all supported signature types for this private key. The ordering is not a comment on the
+    /// preference or security of the contained algorithms.
+    var supportedSignatureAlgorithms: [Certificate.SignatureAlgorithm] { get }
+
+    /// Use the private key to sign the provided bytes with a given signature algorithm.
```

Comment:
How is an async only private key supposed to implement this function? Should it throw or block instead?
Should this be separate protocols for async and sync keys?

some documentation/guidance for conforming types and callers of this function would be good.